### ISSUE TITLE
Fix xdocs Examples AST Consistency Test (Reduce suppressions list)

### ISSUE THREAD

**Issue by smita1078** [checkstyle/checkstyle#18435]
## Overview

The `XdocsExamplesAstConsistencyTest` validates that xdocs example files maintain consistent code structure (differing only in comments and configuration). Currently, we have **228 examples** in the suppression list that need to be reviewed and fixed.

## Goal

Reduce or eliminate entries from the `SUPPRESSED_EXAMPLES` list by fixing the underlying AST issues in example files.

## What Needs to Be Done

Each suppressed example needs investigation to determine:

1. **Is the code actually different?** → Fix the code to match other examples
2. **Should the code be different?** → Mark it as independent (different use case)

## Example: suppresswarningsholder (Already Fixed )

**Before:**
- `SUPPRESSED_EXAMPLES` contained: `Example2`, `Example3`, `Example4`
- These examples were failing validation

**Investigation revealed:**
- Example1 = showcase broad suppression (different/Specific use case)
- Example2, 3, 4 = showcase aliasList variations (should match)

**Fix applied:**
1. Marked Example1 as independent (it has different code intent)
2. Fixed Example2's code to match Example3 and Example4
3. Removed all 3 examples from suppression list kept Example1 as it has specific usecase

**Result:**  Test now passes.

## Current Suppression List (check master HEAD version for latest state)

https://github.com/checkstyle/checkstyle/blob/78f9db1cb555aeb2d37531cb6d49be0448d2c624/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsExamplesAstConsistencyTest.java#L88-L91

### Pick a check from the list above

Example: `checks/annotation/annotationonsameline`

### Investigate the examples

1. Navigate to: `src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/annotation/annotationonsameline/`
2. Look at all `Example*.java` files
3. Read the comments to understand what each example demonstrates
4. Compare the code between examples

### Determine the issue

**Question**: Should these examples have the same code structure or different code structures?

- **Same structure** = They demonstrate the same check with different configuration
- **Different structure** = They demonstrate different use cases of the check

### Apply the fix

**If examples should have the same code:**
- Fix the code so all examples match structurally
- Only comments and configuration values should differ

**If an example legitimately has different code:**
- No code change needed
- The test framework will handle it appropriately, just add in the suppression list.

### Run the test

```bash
mvn clean test -Dtest=XdocsExamplesAstConsistencyTest
```

Verify that your changes resolve the suppression.

### Submit your PR

 **Each check can be fixed independently! Pick one and create one PR, for a single check at a time.**

Example can be found here: https://github.com/checkstyle/checkstyle/commit/040a44c300cd8a692c5ba690f9f3fb0020c5da6b
link to actual issue: #17177

**Comment by ZaheerAhmadDev** [comment]
I am working on  checks/annotation/annotationonsameline/Example2 .

**Comment by SURYANSHUAGRAWAL2006** [comment]
I am on `checks/annotation/suppresswarningsholder/Example1`.

**Comment by SURYANSHUAGRAWAL2006** [comment]
Hello Sir, recently I have solved my check and update example2 to match with all other examples like 3 and 4 except example1 and then I did commit so, the changes will successfully save afterwards I ran the command --->                                                                      `mvn clean test -Dtest=XdocsExamplesAstConsistencyTest` but the error came --->                                                                                              
[INFO] Scanning for projects...
[INFO] Inspecting build with total of 1 modules
[INFO] Installing Central Publishing features
[INFO] 
[INFO] ------------------< com.puppycrawl.tools:checkstyle >-------------------
[INFO] Building checkstyle 12.1.3-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.5.0:clean (default-clean) @ checkstyle ---
[INFO] 
[INFO] --- tidy:1.4.0:check (validate) @ checkstyle ---
[INFO] 
[INFO] --- enforcer:3.6.2:enforce (enforce-maven) @ checkstyle ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] 
[INFO] --- enforcer:3.6.2:enforce (enforce-versions) @ checkstyle ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireJavaVersion passed
[INFO] Rule 1: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] Rule 2: org.apache.maven.enforcer.rules.dependency.DependencyConvergence passed
[INFO] 
[INFO] --- xml:1.2.0:check-format (default) @ checkstyle ---
[INFO] 
[INFO] --- antlr4:4.13.2:antlr4 (default) @ checkstyle ---
[INFO] ANTLR 4: Processing source directory C:\Users\Suryanshu\checkstyle\src\main\resources
[INFO] Processing grammar: com\puppycrawl\tools\checkstyle\grammar\javadoc\JavadocCommentsLexer.g4
[INFO] Processing grammar: com\puppycrawl\tools\checkstyle\grammar\javadoc\JavadocCommentsParser.g4
[INFO] Processing grammar: com\puppycrawl\tools\checkstyle\grammar\java\JavaLanguageLexer.g4
[INFO] Processing grammar: com\puppycrawl\tools\checkstyle\grammar\java\JavaLanguageParser.g4
[INFO] 
[INFO] --- build-helper:3.6.1:add-source (add-source) @ checkstyle ---
[INFO] Source directory: C:\Users\Suryanshu\checkstyle\target\generated-sources\antlr added.
[INFO] 
[INFO] --- resources:3.3.1:resources (default-resources) @ checkstyle ---
[INFO] Copying 400 resources from src\main\resources to target\classes
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-test-source) @ checkstyle ---
[INFO] Skipping plugin execution!
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-it-test-source) @ checkstyle ---
[INFO] Test Source directory: C:\Users\Suryanshu\checkstyle\src\it\java added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-it-test-resource) @ checkstyle ---
[INFO] Skipping plugin execution!
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-xdocs-examples-source) @ checkstyle ---
[INFO] Test Source directory: C:\Users\Suryanshu\checkstyle\src\xdocs-examples\java added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-xdocs-examples-resource) @ checkstyle ---
[INFO] Skipping plugin execution!
[INFO] 
[INFO] --- compiler:3.14.1:compile (default-compile) @ checkstyle ---
[INFO] Recompiling the module because of changed source code.
[INFO] Compiling 470 source files with javac [debug release 17] to target\classes
[INFO] /C:/Users/Suryanshu/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java: Some input files use or override a deprecated API.
[INFO] /C:/Users/Suryanshu/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java: Recompile with -Xlint:deprecation for details.
[INFO] /C:/Users/Suryanshu/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java: C:\Users\Suryanshu\checkstyle\src\main\java\com\puppycrawl\tools\checkstyle\AbstractAutomaticBean.java uses unchecked or unsafe operations.
[INFO] /C:/Users/Suryanshu/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java: Recompile with -Xlint:unchecked for details.
[INFO] 
[INFO] --- plexus-component-metadata:2.2.0:generate-metadata (default) @ checkstyle ---
[INFO] Discovered 9 component descriptor(s)
[INFO] 
[INFO] --- exec:3.6.2:java (default) @ checkstyle ---
[INFO] 
[INFO] --- resources:3.3.1:testResources (default-testResources) @ checkstyle ---
[INFO] Copying 3253 resources from src\test\resources to target\test-classes
[INFO] 
[INFO] --- compiler:3.14.1:testCompile (default-testCompile) @ checkstyle ---
[INFO] Recompiling the module because of changed dependency.
[INFO] Compiling 825 source files with javac [debug release 17] to target\test-classes
[INFO] /C:/Users/Suryanshu/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/DefaultLoggerTest.java: Some input files use or override a deprecated API.
[INFO] /C:/Users/Suryanshu/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/DefaultLoggerTest.java: Recompile with -Xlint:deprecation for details.
[INFO] /C:/Users/Suryanshu/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/AbstractGuiTestSupport.java: Some input files use unchecked or unsafe operations.
[INFO] /C:/Users/Suryanshu/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/AbstractGuiTestSupport.java: Recompile with -Xlint:unchecked for details.
[INFO] 
[INFO] --- jacoco:0.8.14:instrument (default-instrument) @ checkstyle ---
[INFO] 
[INFO] --- surefire:3.5.4:test (default-test) @ checkstyle ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  38.772 s
[INFO] Finished at: 2025-12-28T21:17:43+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.4:test (default-test) on project checkstyle: No tests matching pattern "XdocsExamplesAstConsistencyTest" were executed! (Set -Dsurefire.failIfNoSpecifiedTests=false to ignore this error.) -> [Help 1]                                                                               
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

I checked This error was coming because of `XdocsExamplesAstConsistencyTest` file does not present in `/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsExamplesAstConsistencyTest.java` I did reclone and then check file still does not present  on master branch so, could anyone please help me What is the problem there?
Thank u.

**Comment by romani** [comment]
fetch latest from our repo:
https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsExamplesAstConsistencyTest.java

**Comment by SURYANSHUAGRAWAL2006** [comment]
Hello Sir, I synced with latest upstream master.
Example2/3/4 are already AST-identical and are not present in the suppression list.
Only Example1 remains suppressed as it represents a different use case.
This issue appears to be already resolved in current master, Since I wasn't sending the PR yet so, How this check has automatically updated, at the first time I was updated all examples of this check but unfortunately file was not present so, I set the upstream afterward I saw in checkstyle code base check has already updated so, sir should I pick another check or to be move on some further issues?
Thank u.

**Comment by romani** [comment]
Please choose another Check .
Thanks a lot.

**Comment by smita1078** [comment]
> Example2/3/4 are already AST-identical and are not present in the suppression list. Only Example1 remains suppressed as it represents a different use case. This issue appears to be already resolved in current master, Since I wasn't sending the PR yet so, How this check has automatically updated, at the first time I was updated all examples of this check but unfortunately file was not present so, I set the upstream afterward I saw in checkstyle code base check has already updated so, sir should I pick another check or to be move on some further issues? Thank u.

This was already fixed, as an example for others :  **Example: suppresswarningsholder (Already Fixed )**  it's already mentioned in the issue description.

**Comment by SURYANSHUAGRAWAL2006** [comment]
Hello @smita1078 "I am telling the truth, in example2, class was written, due to this code was differing with example 3 and 4 then I changed it to public class, so that the AST becomes the same or code becomes the same"
Thank u.

**Comment by smita1078** [comment]
> Hello [@smita1078](https://github.com/smita1078) "I am telling the truth, in example2, class was written, due to this code was differing with example 3 and 4 then I changed it to public class, so that the AST becomes the same or code becomes the same" Thank u.

No worries , it looks like your fork might be out of sync with master.  please sync your fork and pull the latest changes.

**Comment by SURYANSHUAGRAWAL2006** [comment]
Hello Sir, I am on `"checks/blocks/rightcurly/Example2"`.

**Comment by seshathri044** [comment]
Hi I am Seshathri
I’m very interested in contributing to this project as part of GSoC. I would love to work on this first issue and learn more about the project. Could you please guide me on how to get started?

Thank you!

**Comment by SURYANSHUAGRAWAL2006** [comment]
Hello @seshathri044 firstly please read check style documentation and contributing guidelines, and try to solve good first issue.
Thank u.

**Comment by SURYANSHUAGRAWAL2006** [comment]
Hello Sir, I Investigate this check and all example files in `"checks/blocks/rightcurly/Example2"` , Example2,Example3,Example4,Example5 have intentionally demonstrate different code structures and these examples are already suppressed in `XdocsExamplesAstConsistencyTest.java` so, conclusion this check already solved.
Sir, could you please suggest me Which next check I picked, please sir.
Thank u.

**Comment by 0xChaitanya** [comment]
@smita1078 I am new to this community. Could you please tell me what is meant by **same check with different configurations**. Can you give an example for the same.

**Comment by smita1078** [comment]
> [@smita1078](https://github.com/smita1078) I am new to this community. Could you please tell me what is meant by **same check with different configurations**. Can you give an example for the same.

Sure,
https://github.com/checkstyle/checkstyle/commit/040a44c300cd8a692c5ba690f9f3fb0020c5da6b here if u see in the example suppresswarningsholder there are 4 examples mentioned under this check.. and 3 of them  i.e(Ex 2,3,4) have same ast structure, where as the 1st example has different ast structure because it has a different use case which can't be modified in order to pass this check as it serves a specific purpose.. therefore here basically we are trying to reduce the size of the list.. and reducing it to have only those example which have a specific use case..

**same check with different configurations** this basically says that those examples which might have changes in the code but we can align the AST structure and they still fulfill their specific purpose.

**Comment by smita1078** [comment]
> I Investigate this check and all example files in `"checks/blocks/rightcurly/Example2"` , Example2,Example3,Example4,Example5 have intentionally demonstrate different code structures and these examples are already suppressed in `XdocsExamplesAstConsistencyTest.java` so, conclusion this check already solved. 

Pick any there are multiple checks like `checks/imports/importorder/
`

**Comment by SURYANSHUAGRAWAL2006** [comment]
Hello Sir,
I analyzed multiple entries from `XdocsExamplesAstConsistencyTest` (e.g.
`SuppressWarningsHolder`, `RightCurly`, `IllegalTokenText`) and in all of them the examples appear to have legitimately different intent, so the current entries in INDEPENDENT_EXAMPLES look correct.
Before continuing further, could you please confirm whether there are still known candidates in this file that need fixing, or if you recommend looking elsewhere for such issues?
Thank u.

**Comment by romani** [comment]
@SURYANSHUAGRAWAL2006 , please make examples same for https://checkstyle.org/checks/annotation/suppresswarningsholder.html#SuppressWarningsHolder
They are already kind of same already but needs final action to make them the same.

**Comment by 0xChaitanya** [comment]
@smita1078, as you raised this issue, you understand this better

Could you please give me feedback on my PR #18544 

My PR didn't reduce the suppression list size, as there were three distinct examples. There were two examples written that were identical, so I removed one of them. However, Example 1 had different code and was not added to the suppression list, so I added that. So, effectively, the size of the suppression list remained unchanged. Is this correct, or am I missing something?

**Comment by smita1078** [comment]
> [@smita1078](https://github.com/smita1078), as you raised this issue, you understand this better
> 
> Could you please give me feedback on my PR [#18544](https://github.com/checkstyle/checkstyle/pull/18544)
> 
> My PR didn't reduce the suppression list size, as there were three distinct examples. There were two examples written that were identical, so I removed one of them. However, Example 1 had different code and was not added to the suppression list, so I added that. So, effectively, the size of the suppression list remained unchanged. Is this correct, or am I missing something?

Since Example3 and Example4 have the same underlying code structure, the goal is to:
Make their code completely identical (same literals, same structure within` // xdoc section -- start/end`)
Keep them demonstrating different features through their different XML configurations:

```
Example3: Default behavior (allowLineBreaks=false)
Example4: With allowLineBreaks=true
```
Remove BOTH from the `SUPPRESSED_EXAMPLES `list

This is the pattern we want to follow for issue #18435 - find examples that can be made consistent, align them, and remove them from suppressions. Each PR should reduce the suppression list size.
If Example1 and 2 truly has a different use case requiring different code structure, that's fine - it can remain in suppressions. However, Example3 and Example4 can definitely be aligned and removed from the list together.

**Comment by ekansh0unofficial** [comment]
I am working on `checks/naming/methodname/ExampleX` . Opening a PR soon

**Comment by SURYANSHUAGRAWAL2006** [comment]
Hello Sir,I want to use Java parsing for AST comparison, as recommended.
Could you please clarify how Java parsing is expected to be used here?
Is there any existing utility, test, or command in Checkstyle to parse xdocs examples and inspect their ASTs?

I tried running the tests locally to inspect ASTs, but I’m hitting execution issues --> Jacoco runtime error <--, so I want to be sure I’m following the intended workflow.
Thank u Sir.

**Comment by romani** [comment]
Just run test in debug, it will not activate jacoco.

**Comment by ayushactiveat** [comment]
I am working on the examples for checks/imports/importorder

**Comment by ishasingh1505** [comment]
I am working on checks/design/onetoplevelclass/Example2.

**Comment by romani** [comment]
Good example of update with links that prove that examples are same https://github.com/checkstyle/checkstyle/pull/18914#issuecomment-3942856125

**Comment by SavalamSamhitha** [comment]
Hello! I am new to contributing to Checkstyle and would like to work on one of the checks for this issue. Could you please guide me if this is still available to work on?

**Comment by romani** [comment]
Yes, still a lot of PRs are expected to be sent

**Comment by Adam-Gamil** [comment]
I am on `checks/coding/illegaltoken/Example2`

**Comment by SavalamSamhitha** [comment]
Thanks for the update. I will work on another example file for this issue.

**Comment by rushita-vachhani** [comment]
Hello @romani @smita1078 ,

I would like to contribute to this issue.

I will work on below issue

`IndentationCheck`

**Comment by abhimarkzz** [comment]
I am working on checks/coding/unnecessaryparentheses/Example2 and Example3.

**Comment by stoyanK7** [comment]
@romani Please reopen issue. List of examples to fix is still long: https://github.com/checkstyle/checkstyle/blob/03125ef0987e8eecab94e6611fc7722fdf3efa31/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsExamplesAstConsistencyTest.java#L81-L90

**Comment by romani** [comment]
Yes, this still open.
GitHub does auto close if uses put "fixes xxxx" in PR description

**Comment by Dhanushreddy09** [comment]
Are there any checks that still needs work @smita1078 ?

**Comment by romani** [comment]
@smita1078 , please look around, in case any leftovers.

But look like this suppression list now serve for use case introduction.
So we can put new issue in javadoc over this list, and we can close this Issue.
Great job , huge effort, enormous archivements!!!!

**Comment by smita1078** [comment]
> Great job , huge effort, enormous archivements!!!!

Thankyou!

> But look like this suppression list now serve for use case introduction. So we can put new issue in javadoc over this list, and we can close this Issue

Yes, The leftover examples are for UseCase Section. Issue is already open https://github.com/checkstyle/checkstyle/issues/19891
```
            // until https://github.com/checkstyle/checkstyle/issues/19891
            "checks/blocks/leftcurly/Example3",
```

> [@smita1078](https://github.com/smita1078) , please look around, in case any leftovers.

Manually I have checked I don't think any is left, once I add validation Test for this https://github.com/checkstyle/checkstyle/issues/20585 We can confirm as well

**Comment by stoyanK7** [comment]
@smita1078 Please send a PR to change the issue number in the Javadoc for `SUPPRESSED_EXAMPLES` to `...19891`. Then we can close this issue. 

Any new violations that pop up under #20585 can be under a separate list/separate `// until ... 20585`

**Comment by smita1078** [comment]
> [@smita1078](https://github.com/smita1078) Please send a PR to change the issue number in the Javadoc for `SUPPRESSED_EXAMPLES` to `...19891`. Then we can close this issue.
    
Sure

> 
> Any new violations that pop up under [#20585](https://github.com/checkstyle/checkstyle/issues/20585) can be under a separate list/separate `// until ... 20585`

Raised PR for this, no new suppression https://github.com/checkstyle/checkstyle/pull/20604

**Comment by smita1078** [comment]
https://github.com/checkstyle/checkstyle/pull/20605

**Comment by romani** [comment]
> Currently, we have 228 examples 

mission impossible is accomplished !!!
thanks a lot to all contributors !!!
